### DISCUSSION TITLE
What's a plugin for you?

### DISCUSSION THREAD

**Discussion by Mattyx00** [fastify/fastify#5337] [Q&A]
Hello guys,
I'm approaching fastify and plugins are very useful but I would like to understand when to use them and when they are not needed.

Two examples:

1. checktoken (file that contains authorize function)

```javascript
const fp = require('fastify-plugin');

async function checkToken(req, res){
    let token = req.headers["x-access-token"] || req.headers["Authorization"] || req.headers["authorization"];
    if (token) {
        //auth control
        const fake_user = {id: 1, name: 'pippo'};
        this.user_model.updateUserById({last_access: new Date()}, fake_user.id);
        req.user = fake_user;
        return;
    }
    return res.status(400).send({ message: "token non valido" });
}

async function plugin(fastify){
    fastify.decorate('authenticate', checkToken);
}

module.exports = fp(plugin);
``` 

2. User queries (file that contains all queries on db about users)
```javascript
const fp = require('fastify-plugin');
const User = require('../../utils/models/User.js');

const queries = {
    getUsers: async () => {
        try {
            let result = await User.query();
            return result;
        } catch (e) {
            console.log(e);
            throw e;
        }
    },
    
    updateUserById: async (payload, idutente) => {
        try {
            const response = await User.query().patchAndFetchById(idutente, payload);
            return response;
        } catch (e) {
            console.log(e);
            throw e;
        }
    },
};

async function plugin(fastify){
    fastify.decorate('user_model', queries);
}

module.exports = fp(plugin);
``` 

Some considerations:
- checktoken as a plugin let me call the function "authenticate" everywhere I need without import the file. That's good.
- user queries as a plugin let me access to this set of db queries everywhere (for example in checktoken I can update user last access without import the queries file)
- that seems so good but now when I access through fastify context I don't see code suggestions about function (parameters and more..), and I can't navigate fast through functions clicking on them
- in more I'm also tempted to make everything a plugin now because it's good having everything everywhere but of course that's not the best way.

So guys, in conclusion, what's for you a plugin? Which example files do you think should be plugins and which ones do you just import?

Thanks to all! <3

**Comment by jessekrubin** [comment]
(Imo) Everything is a plugin (I have written MANY -- some are public if you need a ref).

There's a distinct difference (afaik) between using a plugin wrapped by "fastify-plugin" and one that is not. 
I have one server that connects to hundreds of SQLite dbs. Each is db is NOT wrapped by fp so as to contain each db connection to a subset of routes.

Make anything/everything a plugin but be discerning about if/when you wrap with fastify-plugin. You can always wrap a fp plugin with a non-fp plugin to add a sublayer/context/thingy so that a fp plugin is contained.
